### PR TITLE
Sequence helper idea

### PR DESCRIPTION
**Pull request by stiepan**
Signed-off-by: Kamil Tokarski <ktokarski@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

#### Additional information
- Affected modules and functionalities:

- Key points relevant for the review:

### Tests
- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/operator/sequence_op.h [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+    return expand_;
+  }
+
+  const SampleFrameCtx &GetSampleFrameCtx() {
+    return sample_ctx_;
+  }
+
+  virtual bool ShouldExpandChannels() const {
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (stiepan) [reply dali/pipeline/operator/sequence_op.h:88]: Yeah, false may be more suitable. It was handy default when modifying gaussian_blur.

File: dali/operators/image/convolution/gaussian_blur.cc [hunk 2]
```diff
@@ -62,15 +62,14 @@ The same input can be provided as per-sample tensors.
...
+                         true)
     .AddOptionalArg<float>(kSigmaArgName, "Sigma value for the Gaussian Kernel.",
                            std::vector<float>{0.f}, true)
-    .AddOptionalArg(
-        "dtype", R"code(Output data type.
+    .AddOptionalArg("dtype", R"code(Output data type.
 
 Supported type: `FLOAT`. If not set, the input type is used.)code",
-        DALI_NO_TYPE);
+                    DALI_NO_TYPE);
```
Comment (mzient) [inline dali/operators/image/convolution/gaussian_blur.cc:72]: Please revert - reformatting adds a lot of "changes" to this already large PR, making it even harder to review.

File: dali/pipeline/operator/sequence_op.h [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+  bool IsPerFrame(const TensorVector<CPUBackend> &arg_tensor) {
+    const auto &layout = arg_tensor.GetLayout();
+    return layout.size() > 0 && layout[0] == 'F';
+  }
+
+  template <typename InputBackend>
+  ws_input_t<InputBackend> ExpandInputHelper(const workspace_t<Backend> &ws, int input_idx,
+                                             int num_expand_dims) {
+    static_assert(is_shared_ptr<ws_input_t<InputBackend>>::value,
+                  "Workspace input handle expected to be shared_ptr");
```
Comment (mzient) [inline dali/pipeline/operator/sequence_op.h:230]: This seems overly defensive - after all, there's not much choice of what `workspace_t<InputBackend>` can be.

File: dali/pipeline/operator/sequence_op.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operator/sequence_op.h:88]: That's definitely not a good default.

File: dali/pipeline/operator/sequence_op.h [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+      auto num_expand_dims = expand_desc.NumDims();
+      if (ws.template InputIsType<GPUBackend>(input_idx)) {
+        auto expanded_handle = ExpandInputHelper<GPUBackend>(ws, input_idx, num_expand_dims);
+        AddInputHelper<GPUBackend>(expanded_handle, {ExpandDescFrameInfoFn{&expand_desc}});
+      } else {
+        auto expanded_handle = ExpandInputHelper<CPUBackend>(ws, input_idx, num_expand_dims);
+        AddInputHelper<CPUBackend>(expanded_handle, {ExpandDescFrameInfoFn{&expand_desc}});
+      }
+    }
+  }
```
Comment (mzient) [inline dali/pipeline/operator/sequence_op.h:133]: 1. Do you check that the number of frames matches across the expanded inputs?
2. What happens when one input is expanded and another one isn't? Is there some broadcasting or is it an error?

Comment (stiepan) [reply dali/pipeline/operator/sequence_op.h:133]: 1. I don't validate them. Although, I believe, it is validated by `Operator<Backend>::Setup` call (EnforceUniformInputBatchSize). It seemed fine that the equal number of samples for every input is not enforced by this operator - it doesn't have to rely on that.
2. I deliberately left it unhandled - simply one would be expanded and the other one not -  but thinking again, maybe erroring out by default at the ShouldExpand level would be a better idea - after all it depends on the user input.

File: dali/pipeline/operator/sequence_op.h (hunk 3 above)
Comment (stiepan) [reply dali/pipeline/operator/sequence_op.h:230]: Sure.

**Pull request closed without merging**